### PR TITLE
Add highlight plugin service

### PR DESCRIPTION
**Pull request by waziqi89**
RP-7396
Add Highlighter Plugin Service to support custom highlighter. (Next step: add support of plain highlighter)

General request flow:

1. add highlightFetchTask and feed in the corresponding highlighter from the highlight service
2. highlightFetchTask will check the highlightability of the request at the contrunction time
3. highlightFetchTask will be executed once at the fetch phase during the search execution flow, in which the previous assigned highlighter will be called.

More details:
- We added a new service - highlighter service to initialize the plugins at the startup time and register them, so that later on,  the search request processor can select the corresponding highlighter out of the registry. (The fvh is the only builtin highlighter atm.)
- We also define the highlighter interface and the highlightPlugin interface to provide the list of the custom highlighters. Each highlighter must implement `getHighlights` to do highlighting; `verifyFieldIsSupported` to check the highlightablility for the specific implementation - eg. termvector_order_position for fvh

### CODE REVIEW

File: clientlib/src/main/proto/yelp/nrtsearch/search.proto
Comment (sarthakn7) [inline clientlib/src/main/proto/yelp/nrtsearch/search.proto:823]: Can you please revert these changes? These fields are already in use and modifying them will break the clients. Additionally please follow the [standard protobuf style](https://developers.google.com/protocol-buffers/docs/style) for new fields in this message which specifies using `underscore_separated_names` for field names. We have not followed the standard style in the past but want to do it for new proto objects.

File: src/main/java/com/yelp/nrtsearch/server/grpc/LuceneServer.java
Comment (sarthakn7) [inline src/main/java/com/yelp/nrtsearch/server/grpc/LuceneServer.java:354]: Please add this in alphabetical order

**Review by sarthakn7 — CHANGES_REQUESTED** [review]
I have taken a quick look since this is a draft, but I need a lot more context on these changes. Can you please add it to the PR description? If you can add more details on the changes in the design, what existing classes are being replaced and why that would also be very helpful. Also I think you're referring to ticket RP-7396 not 7395.

File: src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/HighlightFetchTask.java
Comment (sarthakn7) [inline src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/HighlightFetchTask.java:137]: I think this check is specific to fast vector highlighter (Unified does not necessarily need it) and can be moved to its class

File: clientlib/src/main/proto/yelp/nrtsearch/search.proto
Comment (sarthakn7) [inline clientlib/src/main/proto/yelp/nrtsearch/search.proto:846]: Can we have the default behavior as score_ordered?

File: src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/FastVectorHighlighter.java
Comment (sarthakn7) [inline src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/FastVectorHighlighter.java:65]: Can we skip any changes to the behavior of fast vector highlighter as part of this PR? We would need to check how this would affect existing use-cases.
You can add more context in description on any changes in Fast vector highlighter if they are needed as part of this PR.

**Review by sarthakn7 — COMMENTED** [review]
Please also update the documentation for highlighting for any changes to the query or new fields https://github.com/Yelp/nrtsearch/blob/master/docs/highlighting.rst

File: clientlib/src/main/proto/yelp/nrtsearch/search.proto
Comment (waziqi89) [reply clientlib/src/main/proto/yelp/nrtsearch/search.proto:846]: order = none is default in ES, I was trying to match it. However, v1 default is true, v2 default is false introduces confusion to clients. I will change it to be true for both

File: clientlib/src/main/proto/yelp/nrtsearch/search.proto
Comment (waziqi89) [reply clientlib/src/main/proto/yelp/nrtsearch/search.proto:852]: I feel it's more than that. 
highlighter_type is now at the top level, not in settings. (doesn't make sense to be overwrite-able for each filed)
match_field now is a wrapped bool for field to overwrite
Also as we are still under dev, more breaking changes might come in later.

**Comment by waziqi89** [comment]
> Please also update the documentation for highlighting for any changes to the query or new fields https://github.com/Yelp/nrtsearch/blob/master/docs/highlighting.rst

I will update the docs immediately after this PR merged. The PR is not finalized yet, adding docs now is hard.
Additionally, as v1 is still 100% valid after this PR, we can temporarily keep the docs as it is between the PR merge and the read the docs update. This is another + for having all new settings separate in v2.

File: src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/Highlighter.java
Comment (sarthakn7) [reply src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/Highlighter.java:52]: Unless there is some harm in exposing the cache from SearchContext or benefit in isolating it to highlights, I don't see why we should create another context object just to store a map. I also think that by adding to SearchContext it will be available to other plugins and other use-cases as well.

File: src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/FastVectorHighlighter.java
Comment (taoyyu) [inline src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/FastVectorHighlighter.java:40]: I think it might be good to rename this class like FastVectorHighlighterProvider or some name with NRT as prefix to avoid having the same name as org.apache.lucene.search.vectorhighlight.FastVectorHighlighter. But it is totally your call since Elasticsearch is actually using the same name in their repo 😀

File: clientlib/src/main/proto/yelp/nrtsearch/search.proto
Comment (taoyyu) [inline clientlib/src/main/proto/yelp/nrtsearch/search.proto:831]: One question. As a plugin here, do we expect the user to pass custom defined params to the highlighter?

Comment (waziqi89) [reply clientlib/src/main/proto/yelp/nrtsearch/search.proto:831]: In current design, the highlighter is a singleton instance handling all requests. We do not create highlighter per request, so we shouldn't alter it per request. Otherwise it may cause non idempotent issues.

But maybe I am not aware of some use cases where the highlighter must be altered during the execution. Please let me know.

File: src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/FastVectorHighlighter.java
Comment (waziqi89) [reply src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/FastVectorHighlighter.java:40]: Agree. We can add NRT prefix to all builtin highlighters. but the `customHighlighterName`

Comment (waziqi89) [reply src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/FastVectorHighlighter.java:40]: plainHighlighter actually has the same issue. Lucene calls it Highlighter, and ES add prefix"plain" to distinguish them

File: clientlib/src/main/proto/yelp/nrtsearch/search.proto
Comment (taoyyu) [reply clientlib/src/main/proto/yelp/nrtsearch/search.proto:831]: Not so far. Just a random question here since this is an extensible plugin. But settings should be good enough

File: src/main/java/com/yelp/nrtsearch/server/luceneserver/search/SearchContext.java
Comment (sarthakn7) [inline src/main/java/com/yelp/nrtsearch/server/luceneserver/search/SearchContext.java:51]: I suggest renaming this to something like `extraContext` or `cache` or `customContext` or something more general since even though this is called `highlighterContext` it is totally possible for any user of SearchContext to add any custom context in here. Also would be good to add a comment here saying that this is a common context that can be used by plugins otherwise someone reading this code would think that the field is unused and can be removed.

File: clientlib/src/main/proto/yelp/nrtsearch/search.proto
Comment (waziqi89) [reply clientlib/src/main/proto/yelp/nrtsearch/search.proto:823]: It is compatible.  so removing field_match_v2 now

File: src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/HighlightFetchTask.java
Comment (waziqi89) [reply src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/HighlightFetchTask.java:130]: This won't be changed as exceptions can give more information to the users for the failure reason through the error messages.

File: src/main/java/com/yelp/nrtsearch/server/luceneserver/search/SearchContext.java
Comment (waziqi89) [reply src/main/java/com/yelp/nrtsearch/server/luceneserver/search/SearchContext.java:51]: OK. In this case, it shall be non-nullable even if the highlighter task doesn't exists.

File: docs/highlighting.rst
Comment (waziqi89) [reply docs/highlighting.rst:13]: `Requirements` using `-` is a subsection.
`general` using `^` is a subsubsection. general should be under requirements, so I think it's good to keep it as this.

File: docs/highlighting.rst
Comment (waziqi89) [reply docs/highlighting.rst:88]: I was using sublime during the rst development. Will use the tool mentioned here before pushes

File: docs/highlighting.rst
Comment (sarthakn7) [reply docs/highlighting.rst:13]: I meant it should start with upper-case i.e. 
```
General:
^^^^^^^
```

File: src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/HighlightUtils.java
Comment (sarthakn7) [inline src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/HighlightUtils.java:194]: Nit: change to `"Unknown highlighter_type: %s"` so it's formatted similar to most other errors. You can also use just `+` to append the strings if you want

**Review by sarthakn7 — APPROVED** [review]

File: src/main/java/com/yelp/nrtsearch/server/luceneserver/search/SearchRequestProcessor.java
Comment (taoyyu) [inline src/main/java/com/yelp/nrtsearch/server/luceneserver/search/SearchRequestProcessor.java:165]: Is this used already? Do we need to worry about threadsafe for hashmap here now?

Comment (waziqi89) [reply src/main/java/com/yelp/nrtsearch/server/luceneserver/search/SearchRequestProcessor.java:165]: It is not used yet. We want to use it in plan highlighter to cache highlighter per field.
It shall be thread safe though

**Review by taoyyu — APPROVED** [review]

**Pull request merged**
